### PR TITLE
Sampling intermediate RK stages

### PR DESCRIPTION
**Pull request by siuwuncheung**
Implement the capability of sampling intermediate RK stages during offline stage. 

- Command line option `-sample-stages` controls the activation of the new capability.
- New member functions `GetRKStages` and `GetRKTime` are defined in the derived class `HydroODESolver` to enable the sampler access the RK stages and time stored in the time integrator.
- In order to enable the capability, `ode_solver` has to belong to a derived class of `HydroODESolver`, e.g. `RK2AvgSolver` and `RK4ROMSolver`. The new capability is now only compatible with RK2 average and RK4 schemes.

### CODE REVIEW

File: rom/laghos_rom.cpp
Comment (dylan-copeland) [inline rom/laghos_rom.cpp:1286]: I wonder why `make style` did not this fix this already.

File: rom/laghos_timeinteg.cpp
Comment (dylan-copeland) [inline rom/laghos_timeinteg.cpp]: This function should be called only if `!rom`, meaning if not online. Would it be better to do `MFEM_VERIFY(rom)`?

**Comment by dylan-copeland** [comment]
@siuwuncheung After `laghos.cpp:1537` which has `samplerLast = NULL;`, should there be `if (rom_sample_stages) ode_solver_samp->SetSamplerLast(samplerLast);` to set the last sampler to `NULL` in the ODE solver?

File: rom/laghos.cpp
Comment (dylan-copeland) [inline rom/laghos.cpp]: Is there still an issue here?

**Comment by dylan-copeland** [comment]
There are some failing regression tests, e.g. 
Verification failed: (rom_oper) is false:
 --> HydroSolvers expect ROM_Operator.
Also, some space-time regression tests fail, which I don't think is caused by this PR. The problem for those tests is 
Comparing: ROMsol/romS_
ROMsol exists on the baseline branch, but not on the user branch.
triple-point-space-time-romhr: FAIL
Maybe @kevinhkhuynh can help with the `ROMsol` issue.

File: rom/laghos_rom.cpp
Comment (siuwuncheung) [reply rom/laghos_rom.cpp:1286]: This was indeed fixed by `make style`. Probably it was just we hadn't do `make style` before.

File: rom/laghos_utils.cpp
Comment (dylan-copeland) [inline rom/laghos_utils.cpp:358]: The `useXV` and `useVX` options have never been used, and SNS makes them even more obsolete. We probably should remove these options, not necessarily in this PR.

**Review by chldkdtn — APPROVED** [review]

File: rom/laghos_timeinteg.cpp
Comment (siuwuncheung) [reply rom/laghos_timeinteg.cpp]: This routine has been removed as it is no longed needed.

**Comment by chldkdtn** [comment]
@dylan-copeland and @kevinhkhuynh , Let's review and approve this PR today if possible.

**Comment by dylan-copeland** [comment]
@siuwuncheung do the regression tests pass now?

**Comment by kevinhkhuynh** [comment]
> 
> 
> There are some failing regression tests, e.g.
> Verification failed: (rom_oper) is false:
> --> HydroSolvers expect ROM_Operator.
> Also, some space-time regression tests fail, which I don't think is caused by this PR. The problem for those tests is
> Comparing: ROMsol/romS_
> ROMsol exists on the baseline branch, but not on the user branch.
> triple-point-space-time-romhr: FAIL
> Maybe @kevinhkhuynh can help with the `ROMsol` issue.

Are the space-time romhr regression tests working? I have run them manually and no romS files are outputted. Is that intended? If so, then I'll need to update the test script.

**Comment by siuwuncheung** [comment]
@dylan-copeland As @kevinhkhuynh has said, the space-time tests are failing. The `ROMsol` directory for both baseline and user branch in these tests are empty.

File: rom/laghos.cpp
Comment (dylan-copeland) [inline rom/laghos.cpp]: Can this be removed?

File: rom/merge.cpp
Comment (dylan-copeland) [inline rom/merge.cpp]: Is this just for debugging, or is it generally useful? If it should be kept, is it clear what the output is (no label in the output)?

File: rom/laghos_timeinteg.hpp
Comment (dylan-copeland) [inline rom/laghos_timeinteg.hpp]: It seems the return value should be by reference, to avoid copying the data.

File: rom/laghos_timeinteg.cpp
Comment (dylan-copeland) [inline rom/laghos_timeinteg.cpp]: It is unclear whether this would have a significant performance impact, but clearing the vectors and pushing back vectors on each timestep is less efficient than keeping memory allocated throughout the simulation and copying. That is, the constructor for the solver can allocate as many stages as needed in `RKStages`, and then on each stage just copy `S` to `RKStages[i]`.

File: rom/laghos_rom.cpp
Comment (dylan-copeland) [inline rom/laghos_rom.cpp]: Instead of the `typedef` could there just be `std::numeric_limits<double>::max_digits10`?

File: rom/laghos_rom.cpp
Comment (dylan-copeland) [inline rom/laghos_rom.cpp]: opertion -> operation

**Comment by dylan-copeland** [comment]
I confirmed that only the space-time regression tests are failing, and the reason is that the `romS_*` files are not written. These are ROM solutions written to file on each timestep in the spatial ROM case. For space-time, there is only one solution for all time, not solutions at each time. There is no `restore` phase for space-time, yet. When that is implemented, these files at each timestep could be written out to files for comparison.

I am not sure why this fails now but did not before. One solution would be to skip the comparison of these files in the space-time case, but that may be complicated for the regression testing scripts. @kevinhkhuynh do you think this would be easy? We can (i) let the tests fail for now and fix this later, (ii) skip this check for space-time tests (@kevinhkhuynh would need to do that), or (iii) implement a restore phase. For this PR, (i) seems most appropriate, unless (ii) can be done quickly.

**Comment by chldkdtn** [comment]
> I confirmed that only the space-time regression tests are failing, and the reason is that the `romS_*` files are not written. These are ROM solutions written to file on each timestep in the spatial ROM case. For space-time, there is only one solution for all time, not solutions at each time. There is no `restore` phase for space-time, yet. When that is implemented, these files at each timestep could be written out to files for comparison.
> 
> I am not sure why this fails now but did not before. One solution would be to skip the comparison of these files in the space-time case, but that may be complicated for the regression testing scripts. @kevinhkhuynh do you think this would be easy? We can (i) let the tests fail for now and fix this later, (ii) skip this check for space-time tests (@kevinhkhuynh would need to do that), or (iii) implement a restore phase. For this PR, (i) seems most appropriate, unless (ii) can be done quickly.

@dylan-copeland I am okay if we take Option (ii).

**Comment by kevinhkhuynh** [comment]
> 
> 
> I confirmed that only the space-time regression tests are failing, and the reason is that the `romS_*` files are not written. These are ROM solutions written to file on each timestep in the spatial ROM case. For space-time, there is only one solution for all time, not solutions at each time. There is no `restore` phase for space-time, yet. When that is implemented, these files at each timestep could be written out to files for comparison.
> 
> I am not sure why this fails now but did not before. One solution would be to skip the comparison of these files in the space-time case, but that may be complicated for the regression testing scripts. @kevinhkhuynh do you think this would be easy? We can (i) let the tests fail for now and fix this later, (ii) skip this check for space-time tests (@kevinhkhuynh would need to do that), or (iii) implement a restore phase. For this PR, (i) seems most appropriate, unless (ii) can be done quickly.

I just moved the online phase into the offline for now, meaning the online phase still runs, but only the offline stuff is checked.

**Review by kevinhkhuynh — APPROVED** [review]

File: rom/laghos_rom.cpp
Comment (siuwuncheung) [reply rom/laghos_rom.cpp]: Yes it can and should be changed. There is a point when the code does not work and I add this `typedef`, but I later figure out that is not the source of the problem.

File: rom/merge.cpp
Comment (siuwuncheung) [reply rom/merge.cpp]: This is just for debugging. I should remove it.

File: rom/laghos.cpp
Comment (siuwuncheung) [reply rom/laghos.cpp]: Yes it can be safely removed.

File: rom/laghos_rom.cpp
Comment (siuwuncheung) [reply rom/laghos_rom.cpp]: Fixed.

File: rom/laghos.cpp
Comment (dylan-copeland) [inline rom/laghos.cpp:1512]: I'm not sure how the sampler works in libROM (would need to review the code), but I wonder whether it matters to the sampler and its determination of when to sample that multiple samples are taken within the same timestep, with the same timestep size. That is, this is inconsistent input to the sampler. I don't think it matters for Laghos, as we end up sampling every timestep anyway, so we don't actually rely on the sampler to accurately determine when to sample. I am not saying this needs to be changed, just commenting that it is strange and perhaps should be thought about more in the future.

Comment (siuwuncheung) [reply rom/laghos.cpp:1512]: Yes, I think we also have take a similar shortcut by setting an artificial time and timestep in `mergeXV`. Though the code works and every snapshots are sampled, we should review libROM and avoid the potential misuse of `dt`.

**Comment by chldkdtn** [comment]
@siuwuncheung @dylan-copeland is this PR ready to merge?

File: rom/laghos_timeinteg.cpp
Comment (siuwuncheung) [reply rom/laghos_timeinteg.cpp]: Resolved.

**Review by dylan-copeland — APPROVED** [review]

**Comment by siuwuncheung** [comment]
> @siuwuncheung @dylan-copeland is this PR ready to merge?

Now the PR gets all the approval and is ready to merge.

**Pull request merged**
